File: coding-patterns/00-warmup/003.py
# Reverse Vowels (Easy)

class Solution:
  def reverseVowels(self, s: str) -> str:
    # two pointers
    vowels = 'aeiouAEIOU'
    left = 0
    right = len(s)-1
    arr = list(s)

    while left < right:
      while left < right and vowels.find(arr[left]) == -1:
        left += 1
      while left < right and vowels.find(arr[right]) == -1:
        right -= 1
      
      arr[left], arr[right] = arr[right], arr[left]

      left += 1
      right -=1

    return ''.join(arr)

File: coding-patterns/00-warmup/003.cpp
// Reverse Vowels (Easy)

#include <iostream>
#include <string>

class Solution {
  public:
  
    std::string reverseVowels(std::string s) {
      // two pointers

      std::string vowels = "aeiouAEIOU";
      std::string arr = s;

      int start = 0;
      for (int end = s.length()-1; end >= 0; end--) {
          if (start < end && vowels.find(arr[end]) != -1) {
            while (start < end && vowels.find(arr[start]) == -1)
              start++;
            
            std::swap(arr[start++], arr[end]);
          }
      }
      return arr;
    }
};
